stan/math/prim/err/check_bounded.hpp

```diff
@@ -74,6 +74,7 @@ struct bounded<T_y, T_low, T_high, true> {
 template <typename T_y, typename T_low, typename T_high>
 inline void check_bounded(const char* function, const char* name, const T_y& y,
                           const T_low& low, const T_high& high) {
+  STAN_NO_RANGE_AND_SIZE_CHECK;
```

Comment:
This is used to check that real values are in a certain range (like probabilities are bounded [0, 1]) as well as indexes in range.

It looks like this pull is only turning off range checks, but why not value checks too? I think you'd turn them off in similar situations for the same reasons.

Reply:
Oh I forgot this was used for both. Umm I think @syclik had some opinions on this. Also pinging @wds15. I figure this is something users have to explicitly turn on so I'm more in favor of also not doing value checks

Reply:
Value checks? Like what?

Reply:
Not NaN, not infinite, positive definite, valid unit vector, valid simplex -- this sort of thing.

Reply:
that's a different story which one should be able to turn off separately, I think. Something like STAN_FAST_MATH (a synonym for "not safe math, do it fast")

Reply:
If we're allowing for segfaults, then I am also happy allowing for random math explosions. Even though these could happen in really weird, hard to debug ways, it's all better when the checks are flipped back on.

Reply:
Well I guess, how is being unsafe on values practically different than being unsafe on indexing? A lot of the time even when we see these value check errors in practice we ignore them (scale parameter is zero, etc. etc.)

I suppose that these can happen occasionally even in a Stan model we are happy with, and we don't regularly expect to hit out of bounds indexing and stuff.

Reply:
To not stall this unnecessarily, I think it's fine if we do:

1. The macro turns off some but not all index/value checks (this base set seems fine)
2. Only doing the actual index checks for now (`check_bounded` no longer optimized)
3. The macro turns off some but not all index/value checks and we expand it to turn off inf/nan checks
4. Some other variation on a theme

I like one macro better than to two because it would probably just get confusing to have two (for functions like this for instance) and I think the conditions for no index checks are similar to no value checks.

Easy enough to expand the set of functions optimized in production mode at a later time if we only want to do strict indices.

Reply:
Hmmm... that may work... but what is wrong with two macros which do a more tailored thing themselves. To retain easy usability we can introduce things like

```c++
#ifdef NDEBUG
#define MACRO1
#define MACRO2
#endif
```

How does that sound?

Reply:
I like in the future having a `STAN_NDEBUG` which turns off multiple types of checks. Following @t4c1 I changed the name of this macro to `STAN_NO_RANGE_CHECKS` which defines `STAN_NO_RANGE_CHECKS_RETURN return`

Reply:
Are we also not going to remove size checks?

Reply:
> but what is wrong with two macros which do a more tailored thing themselves

I don't see why one would be useful without another -- unsafe is unsafe and it's just hard to remember multiple things.

> not going to remove size checks?

Optimizing out size checks makes sense with this macro I think. `check_bounded` is probly the only one that needs to go if we're not using values.

If we're going for STAN_NDEBUG in the future, does it make sense to just start with that? Or STAN_UNSAFE or STAN_NO_DEBUG since Rok pointed out NDEBUG probably doesn't mean much to people? Then in the future if we decide to split that into multiple things we can and the overall super-unsafe easy-mode flag is still there.

Reply:
I think I'd like to leave the size checks for another time because it changes the exceptions that are returned for some functions like `check_symmetric()` etc.

Reply:
> If we're going for STAN_NDEBUG in the future, does it make sense to just start with that? Or STAN_UNSAFE or STAN_NO_DEBUG since Rok pointed out NDEBUG probably doesn't mean much to people? Then in the future if we decide to split that into multiple things we can and the overall super-unsafe easy-mode flag is still there.

Let's do this in a separate PR once we sort out what that means. Right now I like just having an option to turn off the range checks, then in another PR or an issue we can sort out what we want turned on and off

Reply:
Sounds good